server/http_test.go

```diff
@@ -842,3 +954,477 @@ func TestGetSubscriptionsForChannel(t *testing.T) {
+			request := httptest.NewRequest(http.MethodPut, "/api/v2/subscriptionTemplates", ioutil.NopCloser(bytes.NewBufferString(tc.subscriptionTemplate)))
+			if !tc.skipAuthorize {
+				request.Header.Set(HeaderMattermostUserID, model.NewId())
+			}
+			p.ServeHTTP(&plugin.Context{}, w, request)
+			assert.Equal(t, tc.expectedStatusCode, w.Result().StatusCode)
+		})
+	}
+}
+
+func TestCreateSubscriptionTemplate(t *testing.T) {
+	for name, tc := range map[string]struct {
+		subscriptionTemplate string
+		expectedStatusCode   int
+		skipAuthorize        bool
+		apiCalls             func(*plugintest.API)
+	}{
+		"Invalid": {
+			subscriptionTemplate: "{}",
+			expectedStatusCode:   http.StatusInternalServerError,
+		},
+		"Not Authorized": {
+			subscriptionTemplate: "{}",
+			expectedStatusCode:   http.StatusUnauthorized,
+			skipAuthorize:        true,
+		},
+		"Initial Subscription Template": {
+			subscriptionTemplate: `{"instance_id": "jiraurl1", "name": "some name", "channel_id": "aaaaaaaaaaaaaaaaaaaaaaaaab", "filters": {"events": ["jira:issue_created"], "projects": ["myproject"], "issue_types": ["10001"]}}`,
+			expectedStatusCode:   http.StatusOK,
+			apiCalls: checkHasSubscriptionTemplates([]SubscriptionTemplate{
+				{
+					Filters: getMockSubscriptionFilter("jira:issue_created"),
+				},
+			}, nil, t),
+		},
+		"Initial Subscription Template, GetProject mocked error": {
+			subscriptionTemplate: fmt.Sprintf(`{"instance_id": "jiraurl1", "name": "some name", "channel_id": "aaaaaaaaaaaaaaaaaaaaaaaaab", "filters": {"events": ["jira:issue_created"], "projects": ["%s"], "issue_types": ["10001"]}}`, nonExistantProjectKey),
+			expectedStatusCode:   http.StatusInternalServerError,
+			apiCalls:             hasSubscriptionTemplates([]SubscriptionTemplate{}, t),
+		},
+		"Initial Subscription Template, empty name provided": {
+			subscriptionTemplate: `{"instance_id": "jiraurl1", "name": "", "channel_id": "aaaaaaaaaaaaaaaaaaaaaaaaab", "filters": {"events": ["jira:issue_created"], "projects": ["myproject"], "issue_types": ["10001"]}}`,
+			expectedStatusCode:   http.StatusInternalServerError,
+			apiCalls:             hasSubscriptionTemplates([]SubscriptionTemplate{}, t),
+		},
+		"Initial Subscription Template, long name provided": {
+			subscriptionTemplate: `{"instance_id": "jiraurl1", "name": "` + TestDataLongSubscriptionName + `", "channel_id": "aaaaaaaaaaaaaaaaaaaaaaaaab", "filters": {"events": ["jira:issue_created"], "projects": ["myproject"], "issue_types": ["10001"]}}`,
+			expectedStatusCode:   http.StatusInternalServerError,
+			apiCalls:             hasSubscriptionTemplates([]SubscriptionTemplate{}, t),
+		},
+		"Initial Subscription Template, no project provided": {
+			subscriptionTemplate: `{"instance_id": "jiraurl1", "name": "somename", "channel_id": "aaaaaaaaaaaaaaaaaaaaaaaaab", "filters": {"events": ["jira:issue_created"], "projects": [], "issue_types": ["10001"]}}`,
+			expectedStatusCode:   http.StatusInternalServerError,
```

Comment:
Shouldn't this be StatusBadRequest?

Reply:
https://github.com/Brightscout/mattermost-plugin-jira/pull/23#discussion_r1042953839